### ISSUE TITLE
File I/O support is missing

### ISSUE THREAD

**Issue by JohnLamontagne1996** [WaynePhillipsEA/twinbasic#279]
![1](https://user-images.githubusercontent.com/87338545/126087644-01280420-da40-4cf3-b242-9782cfd6f23d.png)
![2](https://user-images.githubusercontent.com/87338545/126087647-06c6286a-1277-46b6-b4fa-fd3cab58c211.png)
![3](https://user-images.githubusercontent.com/87338545/126087648-665bd329-77a3-4d87-879e-7caa3a7cf000.png)
![4](https://user-images.githubusercontent.com/87338545/126087649-73f88ee3-7555-4ac6-bbdf-3194684e98f0.png)

**Comment by JohnLamontagne1996** [comment]
All of these errors are nonexistent in my Visual Basic 6 application, but it seems for Twinbasic they're problems :(

**Comment by WaynePhillipsEA** [comment]
Thanks for opening the issue.  Indeed, as noted in the preview information page (https://www.twinbasic.com/preview.htm), we haven't implemented any of the file I/O part of the language yet.  That's the only major part of the language that is significantly missing at the moment.  For now please consider using the FSO alternatives.

After we've released the first part of the GUI support over the next couple of weeks, we'll be looking at the file I/O support.

I'll leave this issue open, since I don't think we've got another issue that covers it.

**Comment by KarlSevenSeven** [comment]
The missing file support makes headaches to me.
I can't try some certain DLL code I want to play with without heavily modifying the file parts in it.
The answer "after we support forms" is known.
Before we have file support I can't do some tests, therefore too vague for a preorder.
So my question is when the file stuff will be really available.

**Comment by Kr00l** [comment]
I just want to share my Dir() unicode replacement function which aims to be 100% compatible. Maybe it can help..
```
' (VB-Overwrite)
Public Function Dir(Optional ByVal PathMask As String, Optional ByVal Attributes As VbFileAttribute = vbNormal) As String
Const INVALID_HANDLE_VALUE As Long = (-1)
Const FILE_ATTRIBUTE_NORMAL As Long = &H80
Static hFindFile As Long, AttributesCache As VbFileAttribute
If Attributes = vbVolume Then ' Exact match
    ' If any other attribute is specified, vbVolume is ignored.
    If hFindFile <> 0 Then
        FindClose hFindFile
        hFindFile = 0
    End If
    Dim VolumePathBuffer As String, VolumeNameBuffer As String
    If Len(PathMask) = 0 Then
        VolumeNameBuffer = String$(MAX_PATH, vbNullChar)
        If GetVolumeInformation(0, StrPtr(VolumeNameBuffer), Len(VolumeNameBuffer), ByVal 0, ByVal 0, ByVal 0, 0, 0) <> 0 Then Dir = Left$(VolumeNameBuffer, InStr(VolumeNameBuffer, vbNullChar) - 1)
    Else
        VolumePathBuffer = String$(MAX_PATH, vbNullChar)
        If Left$(PathMask, 2) = "\\" Then PathMask = "UNC\" & Mid$(PathMask, 3)
        If GetVolumePathName(StrPtr("\\?\" & PathMask), StrPtr(VolumePathBuffer), Len(VolumePathBuffer)) <> 0 Then
            VolumePathBuffer = Left$(VolumePathBuffer, InStr(VolumePathBuffer, vbNullChar) - 1)
            VolumeNameBuffer = String$(MAX_PATH, vbNullChar)
            If GetVolumeInformation(StrPtr(VolumePathBuffer), StrPtr(VolumeNameBuffer), Len(VolumeNameBuffer), ByVal 0, ByVal 0, ByVal 0, 0, 0) <> 0 Then Dir = Left$(VolumeNameBuffer, InStr(VolumeNameBuffer, vbNullChar) - 1)
        End If
    End If
Else
    Dim FD As WIN32_FIND_DATA, dwMask As Long
    If Len(PathMask) = 0 Then
        If hFindFile <> 0 Then
            If FindNextFile(hFindFile, FD) = 0 Then
                FindClose hFindFile
                hFindFile = 0
                Exit Function
            End If
        Else
            Err.Raise 5
            Exit Function
        End If
    Else
        If hFindFile <> 0 Then
            FindClose hFindFile
            hFindFile = 0
        End If
        Select Case Right$(PathMask, 1)
            Case "\", ":", "/"
                PathMask = PathMask & "*.*"
        End Select
        AttributesCache = Attributes
        If Left$(PathMask, 2) = "\\" Then PathMask = "UNC\" & Mid$(PathMask, 3)
        hFindFile = FindFirstFile(StrPtr("\\?\" & PathMask), FD)
        If hFindFile = INVALID_HANDLE_VALUE Then
            hFindFile = 0
            If Err.LastDllError > 12 Then Err.Raise 52
            Exit Function
        End If
    End If
    Do
        If FD.dwFileAttributes = FILE_ATTRIBUTE_NORMAL Then
            dwMask = 0 ' Found
        Else
            dwMask = FD.dwFileAttributes And (Not AttributesCache) And &H16
        End If
        If dwMask = 0 Then
            Dir = Left$(FD.lpszFileName(), InStr(FD.lpszFileName(), vbNullChar) - 1)
            If FD.dwFileAttributes And vbDirectory Then
                If Dir <> "." And Dir <> ".." Then Exit Do ' Exclude self and relative path aliases
            Else
                Exit Do
            End If
        End If
        If FindNextFile(hFindFile, FD) = 0 Then
            FindClose hFindFile
            hFindFile = 0
            Exit Do
        End If
    Loop
End If
End Function
```

**Comment by WaynePhillipsEA** [comment]
Fantastic @Kr00l.   I will test it later this week.  Would you allow me to incorporate it into tB?

**Comment by Kr00l** [comment]
Sure @WaynePhillipsEA you can use it.

Something to think about is the FileLen function. It returns in VBx a 'Long'.
The API is..
`LoDWord = GetFileSize(hFile, HiDWord)`

So TB could make a 'FileLen2' function (or whatever name is better) to return a 'LongLong' which joins LoDWord and HiDWord.

**Comment by Kr00l** [comment]
Here another few functions which may help and you can look at/use.

```
' (VB-Overwrite)
Public Function GetAttr(ByVal PathName As String) As VbFileAttribute
Const INVALID_FILE_ATTRIBUTES As Long = (-1)
Const FILE_ATTRIBUTE_NORMAL As Long = &H80
If Left$(PathName, 2) = "\\" Then PathName = "UNC\" & Mid$(PathName, 3)
Dim dwAttributes As Long
dwAttributes = GetFileAttributes(StrPtr("\\?\" & PathName))
If dwAttributes = INVALID_FILE_ATTRIBUTES Then
    Err.Raise 53
ElseIf dwAttributes = FILE_ATTRIBUTE_NORMAL Then
    GetAttr = vbNormal
Else
    GetAttr = dwAttributes
End If
End Function

' (VB-Overwrite)
Public Sub SetAttr(ByVal PathName As String, ByVal Attributes As VbFileAttribute)
Const FILE_ATTRIBUTE_NORMAL As Long = &H80
Dim dwAttributes As Long
If Attributes = vbNormal Then
    dwAttributes = FILE_ATTRIBUTE_NORMAL
Else
    If (Attributes And (vbVolume Or vbDirectory Or vbAlias)) <> 0 Then Err.Raise 5
    dwAttributes = Attributes
End If
If Left$(PathName, 2) = "\\" Then PathName = "UNC\" & Mid$(PathName, 3)
If SetFileAttributes(StrPtr("\\?\" & PathName), dwAttributes) = 0 Then Err.Raise 53
End Sub

' (VB-Overwrite)
Public Sub MkDir(ByVal PathName As String)
If Left$(PathName, 2) = "\\" Then PathName = "UNC\" & Mid$(PathName, 3)
If CreateDirectory(StrPtr("\\?\" & PathName), 0) = 0 Then
    Const ERROR_PATH_NOT_FOUND As Long = 3
    If Err.LastDllError = ERROR_PATH_NOT_FOUND Then
        Err.Raise 76
    Else
        Err.Raise 75
    End If
End If
End Sub

' (VB-Overwrite)
Public Sub RmDir(ByVal PathName As String)
If Left$(PathName, 2) = "\\" Then PathName = "UNC\" & Mid$(PathName, 3)
If RemoveDirectory(StrPtr("\\?\" & PathName)) = 0 Then
    Const ERROR_FILE_NOT_FOUND As Long = 2
    If Err.LastDllError = ERROR_FILE_NOT_FOUND Then
        Err.Raise 76
    Else
        Err.Raise 75
    End If
End If
End Sub
```

**Comment by bclothier** [comment]
FWIW, that is more or less the model that the .NET standard has adopted -- providing the framework itself via the nuget system sot hat it can be maintained and downloaded. I think that is the approach we want to adopt with tB's modern framework.

The legacy syntax should simply call into the modern framework (providing that we are not requiring legacy mode for the project), but a long term goal should be modernizing the codebase as the legacy syntax are frankly obtuse and carries a lot of implicit assumptions that you can't see from reading the source code. Given the requirement to have backward compatibility and the need for legacy mode, we'll still have to provide an implementation that's equivalent, warts and all, to the original VBx libraries. Once that implementation is provided, it probably will never change because it's legacy except to fix issues where the parity is shown to be missing. That would be a separate thing from the modern implementation which should be community-contributed.

**Comment by mansellan** [comment]
@bclothier This is kinda what I was driving at with point 2. in the Additional Contextsection of #111.

When people ask questions about VBA on Stack Overflow, they could actually mean:

1. Actual language features
2. Automatically-referenced type libraries (VBA / VBRUN, stdole)
3. Host object models

For twinBASIC, only @WaynePhillipsEA can influence 1. It's his propriety code, and where the $$$ comes from. 3 is out of scope for twinBASIC (other then making sure that tB is COM-compliant, and works with COM-compliant libraries).

2 Is the interesting one. If I understand correctly, this is what people sometimes call the VB "runtime". It's not a runtime in the traditional sense, rather it's a set of common functions that are always* available for use. These could potentially be dereferenced, and potentially replaced with "modern" equivalents. This should be able to be community-driven.

But @WaynePhillipsEA will need to tell me all the things I said above that are wrong. I'm at the edge of my understanding here.

\* Slight exception for stdole. Although it's automatically added, it can be removed.

**Comment by joyfullservice** [comment]
Just a quick forward-thinking note on the implementation of the file I/O support... Some languages (such as Japanese and Korean) use other characters for path separators, not to mention the backslash vs. forwardslash on Windows vs. Linux/MacOS. Perhaps a `PathSep` function or constant would be a good fit here...

**Comment by KarlSevenSeven** [comment]
I know it's about software, an expected release date is never the release date... ;-)
Nevertheless, what's the current state?

**Comment by dmrvb** [comment]
This is needed for VBx compatibility but for new projects somebody (probably not Wayne) will hopefully produce a class module to do files in a better and more OO way.
I half remember a bug in the VB file system (dunno if it was ever fixed), there is something that it does not do correctly when working with a network drive, this forced me to produce my own file handler. An interesting question....should TB replicate this bug?

**Comment by WaynePhillipsEA** [comment]
Apologies @KarlSevenSeven, I promise I haven't forgotten about this...  this month, hopefully (or at least make a start on it)

**Comment by WaynePhillipsEA** [comment]
> An interesting question....should TB replicate this bug?

Generally my answer to that question would be yes. 
Though if convenient and necessary we should offer an easy switch to fix it.

**Comment by bclothier** [comment]
FWIW, while I agree that we have to replicate the bugs to support legacy codebase that has already has workarounds in place for whatever bugs or explicitly depend on the buggy behavior, I don't think we want a switch. If they are going to update or write a new project, they should be using the current API rather than working with legacy & weird syntax-but-not-really "API". I always thought it was a huge mistake to treat those as statements when they should be objects & methods. Therefore when updating or developing new project, they should be avoiding those legacy syntax altogether.

**Comment by WaynePhillipsEA** [comment]
^ that's a fair point and I agree in this case it makes more sense to steer them away from it (provided we have the modern alternative ready to offer them).

**Comment by Kr00l** [comment]
I would fix it and make the I/O language syntax just correct.
For buggy needs we have #215 Selectable Quirks.

**Comment by KarlSevenSeven** [comment]
> Apologies @KarlSevenSeven, I promise I haven't forgotten about this... this month, hopefully (or at least make a start on it)
I very appreciate your tb project, and I'm very willing to support it.
Perhaps my decision was too early.

The missing file support is a showstopper for me.
Means I can't use tb in a meaninigfull manner.
I have no real project that doesn't do with files.
So I can't report bugs or "all goes well".

Q:
When do you think file support will be available?
If it's this year, that would be nice.

**Comment by WaynePhillipsEA** [comment]
@KarlSevenSeven it will absolutely be started, and hopefully mostly finished, in November.  

I apologise profusely;  family illness caused a slight delay this month, but we're getting back on track.  64-bit is up next, and then it's the file I/O support.

**Comment by WaynePhillipsEA** [comment]
With 64-bit support now in, this is up next.

**Comment by WaynePhillipsEA** [comment]
v0.13.5 is out now, supporting Random and Binary file modes;  Open/Get/Put/Close statements plus library functions FreeFile, EOF, LOF, FileAttr, Loc, Seek, Reset.

Work continues to implement the remaining missing functionality.

**Comment by WaynePhillipsEA** [comment]
I'd like to offer the ability to control the codepage used for reading and writing string data to files.  In particular, this would allow us to provide UTF8 (unicode) support through the file I/O syntax.  The obvious option is to extend the Open statement syntax, though the current syntax already seems a little convoluted:

`Open pathname For mode [ Access access ] [ lock ] As [ # ] filenumber [ Len = reclength ] [CodePage = ?]`

Instead, perhaps we should offer a library function...  `SetFileEncoding(FileNumber As Integer, CodePage As <enum>)`.    Thoughts?

**Comment by XusinboyBekchanov** [comment]
In FreeBasic:

Open filename For Input [encoding_type] [lock_type] As [#]filenumber
Open filename For Output [encoding_type] [lock_type] As [#]filenumber
Open filename For Append [encoding_type] [lock_type] As [#]filenumber

Open filename For Binary [access_type] [lock_type] As [#]filenumber
Open filename For Random [access_type] [lock_type] As [#]filenumber [Len = record_length]


Usage

result = Open( filename[,] For {Input|Output|Append}[,] As filenumber )

or

result = Open( filename[,] For Binary[,] Access {Read|Write}[,] As filenumber )

or

result = Open( filename[,] For Random[,] Access {Read|Write}[,] As filenumber [[,] Len = record_length] )

or

Open filename For {Input|Output|Append} As filenumber

or

Open filename For Binary Access {Read|Write} As filenumber

or

Open filename For Random Access {Read|Write} As filenumber [Len = record_length]

For example:
Open ExePath & "/Tools/Tools.ini" For Output Encoding "utf8" As #Fn

**Comment by bclothier** [comment]
> Thoughts?

As I commented in the linked thread "Selectable Quirks":

> Take file I/Os. For whatever bizarre reasons Microsoft had, they decided to make file I/Os _statements_, not an API. I am not aware of any other languages that makes file I/O a part of the language syntax, but VBx does. That alone brings so many problem -- quick, how do you extend a file open? Oh wait, you can't overload nor shadow it. Do you want Wayne to be the lone maintainer for the file I/O, simply because it's tied up in the compiler & language? I'd rather not. The superior alternative then can be provided via a system similar to nuget (see #16 ), and for special projects, it can be easily extended by overriding or shadowing. 
> ... 
> So for the sake of backward compatibility, it's probably best to treat [File I/O statements] as forever frozen in time to support codebases that contains hacking codes to manipulate the [File I/O statements] in an uncontrolled manner, so that it can port to tB and still work. 

_Originally posted by @bclothier in https://github.com/WaynePhillipsEA/twinbasic/issues/215#issuecomment-945108163_

For the reasons given, I would not want to enhance the existing I/O statements. Any enhancements should be provided via an API that allows us to upgrade or extend the functionality. That should avoid the risk of breaking backward compatibility, especially for the codebases that relies on doing some undocumented wizardry around the File I/O statements.

**Comment by WaynePhillipsEA** [comment]
> > Thoughts?
> 
> As I commented in the linked thread "Selectable Quirks":
> 
> > Take file I/Os. For whatever bizarre reasons Microsoft had, they decided to make file I/Os _statements_, not an API. I am not aware of any other languages that makes file I/O a part of the language syntax, but VBx does. That alone brings so many problem -- quick, how do you extend a file open? Oh wait, you can't overload nor shadow it. Do you want Wayne to be the lone maintainer for the file I/O, simply because it's tied up in the compiler & language? I'd rather not. The superior alternative then can be provided via a system similar to nuget (see #16 ), and for special projects, it can be easily extended by overriding or shadowing.
> > ...
> > So for the sake of backward compatibility, it's probably best to treat [File I/O statements] as forever frozen in time to support codebases that contains hacking codes to manipulate the [File I/O statements] in an uncontrolled manner, so that it can port to tB and still work.
> 
> _Originally posted by @bclothier in [#215 (comment)](https://github.com/WaynePhillipsEA/twinbasic/issues/215#issuecomment-945108163)_
> 
> For the reasons given, I would not want to enhance the existing I/O statements. Any enhancements should be provided via an API that allows us to upgrade or extend the functionality. That should avoid the risk of breaking backward compatibility, especially for the codebases that relies on doing some undocumented wizardry around the File I/O statements.

I disagree on this.  I'm not disagreeing with the general idea that the File I/O syntax should be considered a legacy feature, but we do have to also consider that one of the strong features of tB is its Unicode support, and VBx file I/O is not Unicode compatible.   It would feel very wrong to ignore the Unicode support here, especially when considering that a very minor upgrade would allow for full Unicode support without the developer having to change masses of code to use another API. 

With this proposed feature, the developer would have to opt-in to it (using one of the syntax discussed) and so this absolutely will not conflict with any backwards compatibility in any way, whilst strengthening our overall Unicode support.

**Comment by wqweto** [comment]
Speaking of Unicode, is filename in `Open` statements Unicode? What about `Dir` function?

Do we get Unicode filenames in TB for free (no modification in client code)?

**Comment by WaynePhillipsEA** [comment]
> Speaking of Unicode, is filename in `Open` statements Unicode? What about `Dir` function?
> Do we get Unicode filenames in TB for free (no modification in client code)?

Yes, tB uses CreateFileW().   Dir is not implemented yet.

> Should consider `Encoding "utf-8"` and `Encoding "utf-8 BOM"` to be able to produce/read both for interoperability.

That's a good point; the BOM could help in the case where the developer wants/needs to support both ANSI and Unicode.

**Comment by bclothier** [comment]
> With this proposed feature, the developer would have to opt-in to it (using one of the syntax discussed) and so this absolutely will not conflict with any backwards compatibility in any way, whilst strengthening our overall Unicode support.

What I do want to avoid is adding yet another switch that can affect how a program can behave as was discussed in the linked thread. The more switches there are, the more difficult it will be to maintain and troubleshoot code. We don't want scenarios like this:

```
SetFileCodePage(a)
<several lines of code later>
Open ...
<several more lines>
SetFileCodePage(b)
<several more lines still>
Open ... 'Why isn't this working?!?
```

That might then motivate the programmers to code defensively:

```
SetFileCodePage(a)
Open ...
<several more lines>
SetFileCodePage(a)
Open ...
<several more lines still>
SetFileCodePage(a)
Open ... 
```

Which just adds more noise because we have to be absolutely sure it's the encoding we're intending at that moment. Making it a part of the `Open` statement avoids that problem but then we're now tweaking the legacy statement. 

I hear your arguments about not having to force them to update mass of code to use the new API and such library level functions would make this easy to do. If there was a way to see what option is in effect at the statement -- for example, the CodeLen feature, perhaps -- that would go a long way in minimizing the surprise when a file is not opened in an expected encoding.

**Comment by WaynePhillipsEA** [comment]
> What I do want to avoid is adding yet another switch that can affect how a program can behave as was discussed in the linked thread. The more switches there are, the more difficult it will be to maintain and troubleshoot code. We don't want scenarios like this:
> 
> ```
> SetFileCodePage(a)
> <several lines of code later>
> Open ...
> <several more lines>
> SetFileCodePage(b)
> <several more lines still>
> Open ... 'Why isn't this working?!?
> ```

Yes, now that you've mentioned it, I could absolutely see that one being misused.   So on reflection, I think extending the legacy Open syntax seems more suitable.

**Comment by Kr00l** [comment]
Did some tests with the current available File I/O statements and works good so far.
Default encoding will still stay as ANSI and not utf-16, even when the opt-in is on?

**Comment by mansellan** [comment]
Would the FreeBasic format as posted earlier work @WaynePhillipsEA ? May as well go with prior art if no reason not to?

**Comment by WaynePhillipsEA** [comment]
Yes, matching the FreeBasic syntax seems sensible.

**Comment by WaynePhillipsEA** [comment]
The Print # syntax is now supported in v0.13.6

**Comment by wqweto** [comment]
FreeBasic [doesn't seem to support a lot of encodings](https://documentation.help/FreeBASIC/KeyPgEncoding.html), probably the Linux story is different but on Windows suporting custom codepages (e.g. "cp1251") must be not very complicated.

I recently did [http protocol charset to Windows codepage mapping](https://github.com/wqweto/VbAsyncSocket/blob/master/contrib/cHttpRequest.cls#L1478-L1542) and the sources I used were
- [IANA Character Sets](https://www.iana.org/assignments/character-sets/character-sets.xhtml)
- [Windows Code Page Identifiers](https://docs.microsoft.com/en-us/windows/win32/intl/code-page-identifiers)

**Comment by camomille22003** [comment]
Problem with Seek(),Seek()-x (decrement pointer file Binary access)

![Er_Seek](https://user-images.githubusercontent.com/33807881/143679330-27c0df5a-8ee8-40a7-8a29-2e31a4aac7ed.jpg)
    Public Sub Main()
        Dim F As Integer, x As Integer, i As Integer
        Dim z1 As Byte, z2 As Byte, By As Byte, Bw As Byte
        Dim S As String
        F = FreeFile
        Open "C:\Users\Alain\Desktop\Traducteur\FBdebugger.exe" For Binary Access Read As #F
        x = 1
        Do While Not EOF(F)
          ReDim b(400) As Byte
          i = 0
          Get #F,, z1
          Get #F,, z2
          Get #F,, By
          Get #F,, Bw
          If z1 = 0 And z2 = 0 And (By > &HA1 And By < &HF7) And (Bw > &HA1 And Bw < &HFE) Then
            S = "%" + Hex(By) & "%" + Hex(Bw)
            b(i) = By
            i = i + 1
            b(i) = Bw
            'Hex (Seek(F) - 3)
            Get #F,, By
            Do Until By = 0
              i = i + 1
              If i > 400 Then End
              b(i) = By
              If By < 16 Then S = S + "%0" + Hex(By) Else S = S + "%" + Hex(By)
              Get #F,, By
            Loop
              Debug.Print S
            ReDim Preserve b(0 To i)
            Seek (F), Seek(F) - 1
            x = x + 1
          Else
            Seek (F), Seek(F) - 3
          End If
        Loop
    End Sub

normally the first seek () retrieves the current position and the second subtracts the value and repositioning the pointer.

**Comment by wqweto** [comment]
Try removing the extraneous parentheses so instead of
```
Seek (F), Seek(F) - 3
```
just use
```
Seek F, Seek(F) - 3
```
This is probably a bug with the parser because TB allows instead of
```
Call MyFunction(Param1, Param2)
```
to use the more intuitive 
```
MyFunction(Param1, Param2)
```
which is not allowed in VBx and leads to ambiquos syntax like in your case.

**Comment by WaynePhillipsEA** [comment]
I think Seek is slightly broken.  I'll get it fixed in tomorrow's update.

**Comment by camomille22003** [comment]
same problem with or without parenthes

![Er_Seek_2](https://user-images.githubusercontent.com/33807881/143684243-80370ec6-1b5f-487f-b17a-565228592672.jpg)
es

**Comment by WaynePhillipsEA** [comment]
Apologies, the Seek statement is not yet implemented (but will be tomorrow).  Only the Seek function is currently implemented in v0.13.6

**Comment by JohnLamontagne1996** [comment]
Quick question Wayne, 
Will App.Path be added as well and FileLen, Dir$? Also the two images I attached named Error and Error2 don't seem to occur in my Vb6 project but in TwinBasic
App.Path
![2](https://user-images.githubusercontent.com/87338545/143721129-f5c14d35-eb53-437e-9871-510a7639d83c.png)
Dir$
![3](https://user-images.githubusercontent.com/87338545/143721130-5eeabee3-3695-41c4-9a2e-6fb3d094335f.png)
FileLen
![1](https://user-images.githubusercontent.com/87338545/143721131-d3545450-2837-40ce-aee5-cf2dc78df304.png)
Error
![Error](https://user-images.githubusercontent.com/87338545/143721152-ad2cf910-bdc4-4757-a95a-e1c0c0aeb25e.png)
Error 2
![Error2](https://user-images.githubusercontent.com/87338545/143721153-b28ebd69-85e2-45a6-940e-9e7f64eb2e71.png)

**Comment by WaynePhillipsEA** [comment]
@JohnLamontagne1996 yes, they will all be supported over the next few days as the final parts of the file I/O syntax is added.  Error (1) looks like a bug due to the $ on the call itself.  Error 2 is because file I/O support is not yet complete (Input # / Line Input / Write statements are not yet supported)

**Comment by WaynePhillipsEA** [comment]
v0.13.7 is out now.  Adds support for Seek statement, FileLen, FileDateTime, SetAttr, GetAttr.

**Comment by camomille22003** [comment]
Always as responsive, 
the functioning of seek () is perfectly compatible with VBx. 
Thank you again for all these efforts.

**Comment by Kr00l** [comment]
Any idea how we can get support of `FileLen` and `LOF()` for files > 4 gb ?

**Comment by WaynePhillipsEA** [comment]
@camomille22003  thanks, and you're welcome!

@Kr00l I was just about to open an issue for that :)  We can either change the signature of these function to return a LongLong, or create new functions.  I think changing the signature of the existing functions should not cause any problem... except for when we offer a 'Option Strict' mode that will disallow the LongLong to be downcast to a Long automagically (where existing code assumes the Long return value).

**Comment by Kr00l** [comment]
`Loc() ` is also a limitation.

My feeling says to add new functions and not change the existing signatures. This makes the intent very clear and feels cleaner.

Examples:
```
FileLen64()
LOF64()
Loc64()
```

**Comment by wqweto** [comment]
`Loc` is an utterly useless function which returns the "position of the last byte read or written" in binary mode. But why?

Anyway, probably for compatibility sake it should be implemented too.

Here is one even more obscure function I found hidden in `VBA.Conversion` module.
```
Function MacID(Constant As String) As Long
    Member of VBA.Conversion
    Converts a four-character constant to a Macintosh file type or signature
```
Are we going to get this one too?

**Comment by WaynePhillipsEA** [comment]
I think MacID is already implemented 😜 (though I'm not at my machine to check!)

**Comment by WaynePhillipsEA** [comment]
v0.13.8 is out now, adding support for the `Write #` statement, long file paths and implementations of VBA.FileSystem.MkDir and VBA.FileSystem.RmDir

**Comment by bclothier** [comment]
> @Kr00l I was just about to open an issue for that :) We can either change the signature of these function to return a LongLong, or create new functions. I think changing the signature of the existing functions should not cause any problem... except for when we offer a 'Option Strict' mode that will disallow the LongLong to be downcast to a Long automagically (where existing code assumes the Long return value).

FWIW, in VBA, implicit casting between a `LongLong` and a `Long` will result in a compile-time error of `Type Mismatch`. For that reason, I'm inclined to agree with Kr00l that we shouldn't change the function signature.

**Comment by WaynePhillipsEA** [comment]
@bclothier that's a good catch; I hadn't noticed that.  Perhaps we should also copy that behaviour since it helps prevent LongPtrs being downcast to Long accidently on x64, although it does go against the normal implicit conversion rules.

**Comment by bclothier** [comment]
Yes, I did trip over that when we ported some VBA code used in 64-bit applications to 32-bit and it started complaining about implicit conversions which was a surprising behavior but I think it was deliberate choice by Microsoft to avoid worse outcomes when mixing `Long`s with `LongLong`s esp. when using `LongPtr` which can be one or other depending on the bitness.

**Comment by WaynePhillipsEA** [comment]
v0.13.9 is out now, adding support for `Lock #` and `Unlock #` statements.

**Comment by KarlSevenSeven** [comment]
Played with the file functions; fast and no issue up to now.
The text mode is missing, in order to use Line Input.

**Comment by WaynePhillipsEA** [comment]
Thanks @KarlSevenSeven ... `Line Input #` and `Input #` hopefully coming tomorrow.

**Comment by bclothier** [comment]
Tangential question --- does anyone know why the use of `#` prefix is apparently used but is optional? Last time I checked, it's legal to write with or without the `#` prefix, so it's unclear what the significance if any is attached to the `#` prefix?

**Comment by joyfullservice** [comment]
There is a discussion on this [on StackOverflow](https://stackoverflow.com/questions/13701599/what-does-a-hash-do-to-a-variable-in-vb?answertab=votes#tab-top), but not any real concrete answers. One of those _"it has always worked that way"_ kind of things...  😄  I always thought it was related to [implicit type declarations](https://www.brainbell.com/excel/vba-data-types.html), but those would be added after the variable, not before it.

```vba
?1.5# + 5&
 6.5 
```

**Comment by wqweto** [comment]
Only `Print #nFile, Text` requires `#` to disambiguate from `Print nFile, Text` which print two tab-separated values.

I don't use `#`  anywhere else and frankly don't miss it at all.

**Comment by wopag** [comment]
>>> The Print # syntax is now supported in v0.13.6

TB - Creates the file but doesnt write anything out to it ?		
		Open "c:\orders\a.txt" For Output As #1
		Print #1, "hello"
		Close #1

TB - does not like not having a file path,  sure VB defaults to the current folder ?
              Open "a.txt" For Output As #1
              Print #1, "hello"
              Close #1

**Comment by WaynePhillipsEA** [comment]
@wopag Apologies, a couple of regressions there when I added long path support in v0.13.8.  Now fixed in v0.13.13, thanks.

**Comment by KarlSevenSeven** [comment]
Kill is missing.

**Comment by WaynePhillipsEA** [comment]
@KarlSevenSeven yup, this is still WIP.  Kill / Name-As / Width / Input # / Line Input # / ChDir / ChDrive / FileCopy / CurDir / Dir  are still missing and being worked on.

**Comment by wopag** [comment]
Wayne, do we have a definitive list of  completed or uncompleted commands? 

Be good to know when we do testing that we dont waste your time on commands that are not implemented yet?

**Comment by WaynePhillipsEA** [comment]
@wopag Once the file IO section is complete, the whole VBA standard library should be (mostly) finished and all the language syntax supported (except for GoSub).  I do need to trawl through the docs and type libraries again to check I haven't missed anything though.

Obviously there's still much more to implement from the VB/VBRUN libraries(!), but a lot of that is UI stuff.

**Comment by WaynePhillipsEA** [comment]
v0.13.20 now adds support for the Input # syntax.  From an implementation perspective, the Input # statement is complicated.  As such, there will likely be edge cases that we haven't caught yet.  If you wish to help us out, and you use this feature, please test it let me know if you find any problems.

**Comment by WaynePhillipsEA** [comment]
v0.13.23 now adds ChDir / ChDrive / CurDir$ / CurDir implementations

**Comment by WaynePhillipsEA** [comment]
v0.13.24 now adds FileCopy / Dir / Kill implementations and Name-As syntax support.

**Comment by Kr00l** [comment]
The Kill command does not allow to put a file in the recycle bin. Any idea if/how that aspect can be adressed?

**Comment by WaynePhillipsEA** [comment]
> > v0.13.24 now adds FileCopy / Dir / Kill implementations and Name-As syntax support.
> 
> To check for existence of a file we use :-
> 
> `Debug.Print Dir("filename.txt")`
> 
> Exception error currently - seems if the file doesn't exist it doesn't return a blank string ? ""

Apologies, I was sure we had tests for that.  I'll check and get it fixed tomorrow.  Thanks.

**Comment by WaynePhillipsEA** [comment]
> The Kill command does not allow to put a file in the recycle bin. Any idea if/how that aspect can be adressed?

Looks like the win API in that area is a bit flakey, with SHFileOperation not supporting long paths.

**Comment by Kr00l** [comment]
> > The Kill command does not allow to put a file in the recycle bin. Any idea if/how that aspect can be adressed?
> 
> Looks like the win API in that area is a bit flakey, with SHFileOperation not supporting long paths.

Yes...
Seems only IFileOperation::DeleteFile can address it fully. (Vista+)

**Comment by WaynePhillipsEA** [comment]
> > > v0.13.24 now adds FileCopy / Dir / Kill implementations and Name-As syntax support.
> > 
> > 
> > To check for existence of a file we use :-
> > `Debug.Print Dir("filename.txt")`
> > Exception error currently - seems if the file doesn't exist it doesn't return a blank string ? ""
> 
> Apologies, I was sure we had tests for that. I'll check and get it fixed tomorrow. Thanks.

A rookie error;  reminder to self: On Error Resume Next is the work of the devil.

**Comment by WaynePhillipsEA** [comment]
As of v0.13.25, Line Input # is now supported

**Comment by WaynePhillipsEA** [comment]
As of v0.13.26, Width # is now supported, and so that completes the general File IO support.  Remaining is the proposed encoding extension and support for extra large files, so I will leave this open whilst I address those.

**Comment by WaynePhillipsEA** [comment]
v0.13.27 now adds **experimental** support for the Encoding argument in the Open syntax, as follows:
`Open pathname For mode [Access access] [lock] [Encoding EncString] As [ # ] filenumber [ Len = reclength ]`

EncString can be a string literal / variable.  Allowed values are 

- "default"  (system ANSI codepage, default if not supplied)
- "utf-8"
- "utf-8 bom"
- "utf-16"
- "utf-16 bom"
- any windows codepage number, such as "1252", optionally prefixed with "windows", e.g. "windows-1252"
- plus all the usual suspects, as mentioned by @wqweto [https://github.com/WaynePhillipsEA/twinbasic/issues/279#issuecomment-980526433](url)

UTF16 is mostly supported, but **not** with the Input # statement (you'll get an error), but Line Input # works.  UTF-32 is not supported.

The "utf-8 bom" and "utf-16 bom" versions will output the appropriate BOM file marker when opening a file for output, and will verify the BOM marker when opening a file for input.

**Comment by WaynePhillipsEA** [comment]
I'm closing this issue now, having opened two separate issues for tracking the encoding and big-file support.  For any problems identified with any of the File I/O statements or implementations, please open new issues.  Thanks everyone, for bearing with us.
